crates/router/src/services/authentication/decision.rs

```diff
@@ -0,0 +1,249 @@
+use common_utils::{errors::CustomResult, request::RequestContent};
+use masking::Secret;
+use router_env::opentelemetry::KeyValue;
+use serde::Serialize;
+use storage_impl::errors::ApiClientError;
+
+use crate::{core::metrics, routes::SessionState};
+
+// # Consts
+//
+
+const DECISION_ENDPOINT: &str = "/rule";
+const RULE_ADD_METHOD: common_utils::request::Method = common_utils::request::Method::Post;
+const RULE_DELETE_METHOD: common_utils::request::Method = common_utils::request::Method::Delete;
+
+pub const REVOKE: &str = "REVOKE";
+pub const ADD: &str = "ADD";
+
+// # Types
+//
+
+/// [`RuleRequest`] is a request body used to register a new authentication method in the proxy.
+#[derive(Debug, Serialize)]
+pub struct RuleRequest {
+    /// [`tag`] similar to a partition key, which can be used by the decision service to tag rules
+    /// by partitioning identifiers. (e.g. `tenant_id`)
+    pub tag: String,
+    /// [`variant`] is the type of authentication method to be registered.
+    #[serde(flatten)]
+    pub variant: AuthRuleType,
+    /// [`expiry`] is the time **in seconds** after which the rule should be removed
+    pub expiry: Option<u64>,
+}
+
+#[derive(Debug, Serialize)]
+pub struct RuleDeleteRequest {
+    pub tag: String,
+    #[serde(flatten)]
+    pub variant: AuthType,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum AuthType {
+    /// [`ApiKey`] is an authentication method that uses an API key. This is used with [`ApiKey`]
+    ApiKey { api_key: Secret<String> },
+}
+
+#[derive(Debug, Serialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum AuthRuleType {
+    /// [`ApiKey`] is an authentication method that uses an API key. This is used with [`ApiKey`]
+    /// and [`PublishableKey`] authentication methods.
+    ApiKey {
+        api_key: Secret<String>,
+        identifiers: Identifiers,
+    },
+}
+
+#[allow(clippy::enum_variant_names)]
+#[derive(Debug, Serialize, Clone)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum Identifiers {
+    /// [`ApiKey`] is an authentication method that uses an API key. This is used with [`ApiKey`]
+    ApiKey { merchant_id: String, key_id: String },
+    /// [`PublishableKey`] is an authentication method that uses a publishable key. This is used with [`PublishableKey`]
+    PublishableKey { merchant_id: String },
+}
+
+// # Decision Service
+//
+
+pub async fn add_api_key(
+    state: &SessionState,
```

Comment:
@akshay-97 , we cannot pass standalone `ApiClient` as the `send_request` function takes `SessionState` as an input argument